Consider Container Percentage Sizes When Determining Definite Cross Size

sammygill opened the pull request:
#### 39cc41965e9ea32db1eecab5bea00b539397d22a
<pre>
Consider Container Percentage Sizes When Determining Definite Cross Size
<a href="https://bugs.webkit.org/show_bug.cgi?id=245194">https://bugs.webkit.org/show_bug.cgi?id=245194</a>
rdar://99469852

Reviewed by Alan Bujtas.

When computing the transferred size suggestion for the minimum size of
flex items, we need to use the intrinsic aspect ratio and the definite
cross size. Both the Flexbox specification and CSS-Sizing specification
provide information on when a size can be considered definite. This
patch implements the following portion of the Flexbox spec to determine
the definite cross size:

If a single-line flex container has a definite cross size, the outer
cross size of any stretched flex items is the flex container’s inner
cross size (clamped to the flex item’s min and max cross size) and is
considered definite.

The new test case provides an instance of a scenario where this addition
is important. The image is being used as a flex item, but the flex
container is being constrained by another containing block that has a
specified height that is smaller than the image. When computing the
automatic minimum size (specifically the transferred size suggestion),
we can determine that the flex container has a definite cross size using
its percentage specified height along with the height available to it.
Then, we can use this cross size along with the image&apos;s aspect ratio
to determine the appropriate main size.

Spec: <a href="https://www.w3.org/TR/css-flexbox-1/#min-size-auto">https://www.w3.org/TR/css-flexbox-1/#min-size-auto</a>
<a href="https://www.w3.org/TR/css-flexbox-1/#definite">https://www.w3.org/TR/css-flexbox-1/#definite</a>

* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::RenderFlexibleBox::childCrossSizeShouldUseContainerCrossSize const):
(WebCore::RenderFlexibleBox::computeCrossSizeForChildUsingContainerCrossSize const):

Canonical link: <a href="https://commits.webkit.org/254758@main">https://commits.webkit.org/254758@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c6e15d782280682b3858cfc466d640718de3c4be

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/90043 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/34590 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/20679 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/99377 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/156882 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/94049 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/33082 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/28457 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/82380 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/93673 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/95691 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/26292 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/76854 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/26205 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/81152 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/80952 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/69206 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/66/builds/34282 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/15034 "Passed tests") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/67/builds/32123 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/15975 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3341 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/63/builds/35867 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/38935 "Passed tests") | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/65/builds/37768 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/35059 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

rwlbuis commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -1826,8 +1825,11 @@ LayoutUnit RenderFlexibleBox::computeCrossSizeForChildUsingContainerCrossSize(co
         // Let's compute the definite size value for the flex item (value that we can resolve without running layout).
         auto isHorizontal = isHorizontalFlow();
         auto size = isHorizontal ? style().height() : style().width();
-        ASSERT(size.isFixed());
```
Why can this ASSERT be removed and not added to?

sammygill replied:
Actually I forgot that these are not in release builds. Initially I thought it was a little much too for the synchronization between this method and childCrossSizeShouldUseContainerCrossSize as we implement more of the definite sizes, but now I realize it's probably fine. Will add to it.

cathiechen commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -1826,8 +1825,11 @@ LayoutUnit RenderFlexibleBox::computeCrossSizeForChildUsingContainerCrossSize(co
+        if (std::optional<LayoutUnit> percentageBasedSize; size.isPercent() && (percentageBasedSize = containingBlockLogicalHeightForContent(AvailableLogicalHeightType::IncludeMarginBorderPadding)))
+            definiteValue = percentageBasedSize.value();
+        else
+            definiteValue = size.value();
```
Looks like the `else` is not needed?

sammygill replied:
Ah yes you are right. I think what I was supposed to do was change definiteValue so that it was only declared and not defined above. The code then would have looked something like this: 

        auto definiteValue;
        if (std::optional<LayoutUnit> percentageBasedSize; size.isPercent() && (percentageBasedSize = containingBlockLogicalHeightForContent(AvailableLogicalHeightType::IncludeMarginBorderPadding)))
            definiteValue = percentageBasedSize.value();
        else
            definiteValue = size.value();

Do you think one version is better than the other or have a preference? One benefit of this would be that definiteValue only gets written once but I think either will work.

rwlbuis replied:
I think Cathie is right and looks like you can avoid the percentageBasedSize temp var as well.

sammygill replied:
Ok I will change that logic. So do you think the check for size.isPercent() is sufficient for that if clause then? what happens if containingBlockLogicalHeightForContent returns an optional that is empty/null?

rwlbuis replied:
In cases like this usually value_or is used.

sammygill replied:
Ok the changes have been made. I also realized containingBlockLogicalHeightForContent was the wrong method to use here. That was setting the cross size based off of the total available logical height, but we need it based off the % here. I think availableLogicalHeightForPercentageComputation is the right method to use here. I'll also be updating the test case to make sure it covers this.

rwlbuis replied:
Patch looks good but I wonder in what cases availableLogicalHeightForPercentageComputation can return null? And also if null is returned whether using 0_ul is the right thing to do.

rwlbuis replied:
Looking at this again, I think min-width/height and max-width/height could use the same treatment (support percentage), but I think it is enough to leave a FIXME for that.

alanbaradlay commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -1020,9 +1020,8 @@ bool RenderFlexibleBox::childCrossSizeShouldUseContainerCrossSize(const RenderBo
         auto& crossSize = isHorizontalFlow() ? style().height() : style().width();
-        return crossSize.isFixed();
+        return crossSize.isFixed() || (crossSize.isPercent() && availableLogicalHeightForPercentageComputation()); 
     }
```
This could probably be just:
return hasDefiniteLogicalHeight();

I am a bit concerned about the 0px cross size value we are going to return when we fail to resolve the definite value. It could produce a collapsed flex item.

We could do something like this:

```
std::optional<LayoutUnit> RenderFlexibleBox::computeCrossSizeForChildUsingContainerCrossSize(const RenderBox& child) const
{
    if (crossAxisIsPhysicalWidth())
        return contentWidth();

    // Let's compute the definite size value for the flex item (value that we can resolve without running layout).
    auto isHorizontal = isHorizontalFlow();
    auto resolvedValue = [&] () -> std::optional<LayoutUnit> {
        auto size = isHorizontal ? style().height() : style().width();
        if (size.isFixed())
            return { LayoutUnit { size.value() } };
        if (size.isPercent()) {
            auto resolvedPercentValue = availableLogicalHeightForPercentageComputation();
            // childCrossSizeShouldUseContainerCrossSize said we can resolve percent values.
            ASSERT(resolvedPercentValue);
            return resolvedPercentValue;
        }
        ASSERT_NOT_REACHED();
        return { };
    }();

    if (!resolvedValue)
        return { };
    // Keep this sync'ed with childCrossSizeShouldUseContainerCrossSize().
    auto definiteSizeValue = [&] {
        auto definiteValue = *resolvedValue;
        auto maximumSize = isHorizontal ? style().maxHeight() : style().maxWidth();
        if (maximumSize.isFixed())
            definiteValue = std::min(*resolvedValue, LayoutUnit { maximumSize.value() });

        auto minimumSize = isHorizontal ? style().minHeight() : style().minWidth();
        if (minimumSize.isFixed())
            definiteValue = std::max(*resolvedValue, LayoutUnit { minimumSize.value() });

        return definiteValue;
    };
    return { std::max(0_lu, definiteSizeValue() - crossAxisMarginExtentForChild(child)) };
}  
```
However it only helps with one of the 2 callsites:
 _computeChildIntrinsicLogicalWidths()_  -> could fallback to an auto override size scope (same behavior as before)
 _computeMainSizeFromAspectRatioUsing()_ ->  needs a way to return a sane value (the safest would be to propagate the optional value all the way back to _computeFlexBaseSizeForChild()_ -a rather verbose change)

Or, instead of a band-aid, we could address the actual issue and let _childCrossSizeShouldUseContainerCrossSize()_ return the  definite value that we can somehow pass it in to _computeCrossSizeForChildUsingContainerCrossSize()_  (another verbose change). 
It would certainly eliminate the uncertainty introduced by this multi layer structure where the decision point and the place where we try to resolve the value are so disconnected from each other.

alanbaradlay commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -1020,9 +1020,8 @@ bool RenderFlexibleBox::childCrossSizeShouldUseContainerCrossSize(const RenderBo
-        // FIXME: so far we're only considered fixed sizes but we should extend it to other definite sizes.
         auto& crossSize = isHorizontalFlow() ? style().height() : style().width();
-        return crossSize.isFixed();
+        return crossSize.isFixed() || hasDefiniteLogicalHeight(); 
```
fixed logical height value should already be covered by hasDefiniteLogicalHeight()

alanbaradlay commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -231,9 +232,9 @@ void RenderFlexibleBox::computeChildIntrinsicLogicalWidths(RenderObject& childOb
     // the intrinsic sizes are properly computed in the presence of aspect ratios. The only exception is when
     // we are both a flex item&container, because our parent might have already set our overriding size.
-    if (childCrossSizeShouldUseContainerCrossSize(child) && !isFlexItem()) {
+    if (std::optional<LayoutUnit> crossSize = childCrossSizeShouldUseContainerCrossSize(child); crossSize && !isFlexItem()) {
```
this could be just auto

alanbaradlay commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -925,8 +926,9 @@ LayoutUnit RenderFlexibleBox::computeMainSizeFromAspectRatioUsing(const RenderBo
-        crossSize = computeCrossSizeForChildUsingContainerCrossSize(child);
+        std::optional<LayoutUnit> containerCrossSize = childCrossSizeShouldUseContainerCrossSize(child); 
+        ASSERT(containerCrossSize);
+        crossSize = computeCrossSizeForChildUsingContainerCrossSize(child, containerCrossSize);
```
This is still a bit problematic as we could very well end up with a std:null_opt here and it's too late to fallback to a non-zero height.

alanbaradlay commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -1019,10 +1019,7 @@ bool RenderFlexibleBox::childCrossSizeShouldUseContainerCrossSize(const RenderBo
-        // FIXME: so far we're only considered fixed sizes but we should extend it to other definite sizes.
-        auto& crossSize = isHorizontalFlow() ? style().height() : style().width();
-        return crossSize.isFixed();
+        return hasDefiniteLogicalHeight(); 
```
_hasDefiniteLogicalHeight_ calls _availableLogicalHeightForPercentageComputation_ which has this odd early return on _skipContainingBlockForPercentHeightCalculation_.  It makes me a bit uneasy and it may return false for fixed width box. We should probably go back to the original approach here.

sammygill committed:
Consider Container Percentage Sizes When Determining Definite Cross Size
https://bugs.webkit.org/show_bug.cgi?id=245194
rdar://99469852

Reviewed by Alan Bujtas.

When computing the transferred size suggestion for the minimum size of
flex items, we need to use the intrinsic aspect ratio and the definite
cross size. Both the Flexbox specification and CSS-Sizing specification
provide information on when a size can be considered definite. This
patch implements the following portion of the Flexbox spec to determine
the definite cross size:

If a single-line flex container has a definite cross size, the outer
cross size of any stretched flex items is the flex container’s inner
cross size (clamped to the flex item’s min and max cross size) and is
considered definite.

The new test case provides an instance of a scenario where this addition
is important. The image is being used as a flex item, but the flex
container is being constrained by another containing block that has a
specified height that is smaller than the image. When computing the
automatic minimum size (specifically the transferred size suggestion),
we can determine that the flex container has a definite cross size using
its percentage specified height along with the height available to it.
Then, we can use this cross size along with the image's aspect ratio
to determine the appropriate main size.

Spec: https://www.w3.org/TR/css-flexbox-1/#min-size-auto
https://www.w3.org/TR/css-flexbox-1/#definite

* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::RenderFlexibleBox::childCrossSizeShouldUseContainerCrossSize const):
(WebCore::RenderFlexibleBox::computeCrossSizeForChildUsingContainerCrossSize const):

Canonical link: https://commits.webkit.org/254758@main
